### DISCUSSION TITLE
Bumping prerelease dependencies with poetry

### DISCUSSION THREAD

**Discussion by jvansanten** [renovatebot/renovate#11685] [Request Help]
I'm currently evaluating switching from GitHub-native Dependabot to Renovate for a suite of coupled Python packages. Each of these has 3 general groups of dependencies (all from pypi), which I would like to treat differently:

- external dependencies: update ~monthly, bundled into single PR
- dev dependencies: ditto
- internal dependencies: update as rapidly as possible, including prereleases

I'm having trouble triggering the 3rd group, but am unsure if that's because Renovate doesn't support what I have in mind, I haven't set the right config options, or I have fundamentally misunderstood how this is supposed to work. I have this config:

https://github.com/AmpelProject/Ampel-alerts/blob/107b5369a70c020c89bee985fff9928aaa3e879f/renovate.json

In the logs of the hosted app I see:

```
DEBUG: packageFiles with updates
...
          {
            "depName": "ampel-interface",
            "depType": "dependencies",
            "currentValue": "^0.8.0-alpha.0",
            "managerData": {
              "nestedVersion": true
            },
            "datasource": "pypi",
            "lockedVersion": "0.8.0a6",
            "versioning": "poetry",
            "depIndex": 3,
            "warnings": [],
            "sourceUrl": "https://github.com/AmpelProject/Ampel-interface",
            "homepage": "https://ampelproject.github.io",
            "updates": []
          },
```

I would like Renovate to bump `currentValue` from `^0.8.0-alpha.0` to `^0.8.0-alpha.7`, the latest prerelease. Running locally with trace logging, I see:

```
TRACE: ampel-interface matches against /^ampel-.*/u (repository=AmpelProject/Ampel-alerts)
TRACE: mergeChildConfig (repository=AmpelProject/Ampel-alerts)
<big config dump>
TRACE: currentValue ^0.8.0-alpha.0 is range (repository=AmpelProject/Ampel-alerts)
TRACE: currentValue ^0.8.0-alpha.0 is range (repository=AmpelProject/Ampel-alerts)
TRACE: mergeChildConfig (repository=AmpelProject/Ampel-alerts)
<another big config dump>
```

which seems to indicate that it has ended up here, https://github.com/renovatebot/renovate/blob/dfef9f3cb130ddbde0077129367c460912416a7c/lib/workers/repository/process/lookup/current.ts#L24, but I can't figure out where new versions are queried from pypi and compared to the spec in pyproject.toml. Can anyone give me a pointer on how to investigate further?

**Comment by jvansanten** [comment]
Answering a bit of my own question: the primary problem here is the way that `poetry` and PyPI's versioning schemes interact. While `poetry` uses SemVer in its package specification, internally it takes a sort of union of SemVer and PEP440, which it then emits into artifacts uploaded to PyPI as PEP440:

https://github.com/python-poetry/poetry-core/blob/01c0472d9cef3e1a4958364122dd10358a9bd719/poetry/core/version/version.py#L19-L73

For example, the version "0.8.0-alpha.0" in pyproject.toml gets published to PyPI as "0.8.0a0".

`renovate`'s implementation of `poetry` versioning only treats the SemVer style, and uses its validity check to filter the list of versions returned from PyPI queries. For point releases SemVer and PEP440 are identical, so everything works. For alpha releases, however, the poetry versioning scheme rejects versions like "0.8.0a0". The cleanest solution here is probably to make the poetry versioning scheme accept either SemVer or PEP440, but I'm not sure how large of a change that would be.

**Reply by rarkins** [reply]
We do already have a hacked-together "union" of semver and pep440 in our poetry versioning, and aware it needs improvement, see https://github.com/renovatebot/renovate/issues/9914

At least two of us have attempted quick fixes but always managed to cause other things to fail and decided to abort the attempt.

In short: you're right that we struggle a bit when it comes to how Poetry combines semver and pep440
